File: leetcode_1217.java
/*

1217. Minimum Cost to Move Chips to The Same Position
URL = https://leetcode.com/problems/minimum-cost-to-move-chips-to-the-same-position/

THOUGHT PROCESSES : 

[T,S] = [O(n), O(1)] ideally - linear scan approach - where n denotes number of chips


Edge cases
Can have a position of length = 1 only
Can have all array values be the same too
Each position is guaranteed a chip : minCostToMove is guaranteed to be bounded by [0,x), in the integers

Shifting by 2 = no cost. Optimize for this first!
Shifting by 1 = incur a cost. Do not optimize for this first??


First idea : greedy scanning in an array checks for neighbors
E.g. at A[i], check your neighbors A[i+1], A[i+2]

Track for max number of chips in a position too?


Easy to intuit when dealing with a subarray of size = 2 here!


[1,1,1,3,3] - min cost = 0; no matter movement
[2,2,2,3,3] - min cost = 2, but how to know to move 3rd_pos to 2nd_pos here?
-> In this case, we know a move is neccessary, so we take the min of each position, and based on that, perform the increment of minNumMoves needed

Array length may also be known ahead-of-time too : find out parity - is it odd length or even length?

In some greedy approaches, pre-computation and scanning an entire list of items ie needed
Greedy is used predominantly in graph-walking



*/


class Solution {
    public int minCostToMoveChips(int[] position) {
        
    }
}

File: leetcode_1217.go
/*
URL := https://leetcode.com/problems/minimum-cost-to-move-chips-to-the-same-numChips/
1217. Minimum Cost to Move Chips to The Same numChips

*/
import (
    "math"
)

func minCostToMoveChips(numChips []int) int {
    oddSumCount := 0
    evenSumCount := 0
    // [-1] Calculate the number of chips for each given position too
    // numChips := []int{}
    // for x := 0; x < n; x++ {
    //     numChips = append(numChips, 0)
    // }

    // Your approach is right : just add if even or if odd here ( the test cases )
    // Utilize a hashmap instead of an array
    // Ok we can have a huge number of positions, but should not run into TLE here
    // [0] Store of counts for minChip costs
    leftChipCountMap := make(map[int]int)
    rightChipCountMap := make(map[int]int)
    // [1] March rightwards
    n := len(numChips)
    sort.Ints(numChips) // asc order too
    for i := 0; i < n; i++ {
        if numChips[i] % 2 == 1 { // odds
            oddSumCount += 1
            if _, ok := leftChipCountMap[numChips[i]]; ! ok {
                leftChipCountMap[numChips[i]] = evenSumCount
            }
        } else if numChips[i] % 2 == 0 { // hit an event index ( after an odd hit )
            evenSumCount += 1
            if _, ok := leftChipCountMap[numChips[i]]; ! ok {
                leftChipCountMap[numChips[i]] = oddSumCount
            }
        }
    }
    // [2] March leftwards
    oddSumCount = 0
    evenSumCount = 0
    for i := n-1; i >= 0; i-- {
        if numChips[i] % 2 == 0 { // evens
            evenSumCount += 1
            if _, ok := rightChipCountMap[numChips[i]]; ! ok {
                rightChipCountMap[numChips[i]] = oddSumCount
            }
        } else if numChips[i] % 2 == 1  {
            oddSumCount += 1
            if _, ok := rightChipCountMap[numChips[i]]; ! ok {
                rightChipCountMap[numChips[i]] = evenSumCount
            }
        }
    }
    // [3] Get global min
    myMinCost := math.MaxInt32
    // for i := 0; i < len(numChips); i++ {
    for k, _ := range leftChipCountMap {
        // Paran enclosures : ensure that statements terminate
        myMinCost = (int)(math.Min((float64)(myMinCost),(float64)(leftChipCountMap[k] + rightChipCountMap[k])))
    }
    return myMinCost
}
